### DISCUSSION TITLE
Migrating from integer primary key to uuid

### DISCUSSION THREAD

**Discussion by aaronsgithub** [sqlalchemy/sqlalchemy#10405] [Usage Questions]
The database I'm working on isn't in production, but my goal is to learn what I would do if it was.
My apologies if a prior discussion exists somewhere on this topic, but I could not find one.

For the sake of the below discussion, I'm using the ORM and alembic for migrations, with a PostgreSQL backend.

Suppose we have an existing user model / table which uses an integer primary key:

```
class User(Base):
    __tablename__ = "user"
    id: Mapped[int] = mapped_column(primary_key=True)
    # other fields here

    profile: Mapped["Profile"] = relationship(back_populates="user")
    
class Profile(Base):
    __tablename__ = "profile"
    id: Mapped[int] = mapped_column(ForeignKey("user.id", ondelete="CASCADE"), primary_key=True)
    profile: Mapped["Profile"] = relationship(back_populates="profile") 
    
# and some other models with foreign keys to User
```

Now suppose we want to change our id on User / user to UUIDs, so that our id column might end up looking like:
```
    id: Mapped[UUID] = mapped_column(sa.UUID, primary_key=True, server_default=func.uuid_generate_v4())
```

What are some best practices, in general as well as SQLAlchemy specific, to approach this?

My idea at the moment consists of the following steps:

- create the uuid column e.g User.uuid
- make the uuid column the primary key
- rename the id column to old_id
- rename the uuid column to id
- delete the old_id column
- update any foreign keys using the old_id column to use the id column

What steps am I missing?
And how should these steps be ordered?

For any foreign key which points to the existing User.id field, should I create a new foreign key pointing to User.uuid, and then delete the old foreign keys?

I imagine that in a production setting, the process might be split up over multiple deployments / migrations.

**Comment by CaselIT** [comment]
Hi,

If the column to migrate has foreign keys associated I think what you have to do as you suggested.
If it does not have any you may be able to update directly the type using the `using expression` part of alter column available on postgresql. https://www.postgresql.org/docs/current/sql-altertable.html

**Comment by zzzeek** [comment]
if you turn uuid into the primary key, you can then just drop the "old id" column directly.   the tricky part here would be updating the FK references, since different DBs will require you to do this atomically, like PostgreSQL.   on PG, if you change the old id to not be a PK col anymore, it wont allow that with the existing FK references.

basically how to do this is pretty DB specific so the which DB backend detail is pretty important here.

**Reply by CaselIT** [reply]
> on PG, if you change the old id to not be a PK col anymore, it wont allow that with the existing FK references.

If i'm not mistaken an unique index/constraint is enough to allow fk

**Reply by zzzeek** [reply]
this is true, so they'd have to create this constraint on the old id column first if they want to make it no longer the PK.

if this is PostgreSQL, I think using PG's constructs to change the PK and update the values directly would be the way to go.

**Reply by zzzeek** [reply]
all of these operations too are subject to long delays if this is a large table with a lot of traffic
